TailLight/wmi.cpp

```diff
@@ -1,6 +1,93 @@
 #include "driver.h"
 
 
+
+VOID FakeBISTTimerProc(_In_ WDFTIMER timer) {
+    WDFDEVICE Device = static_cast<WDFDEVICE>(WdfTimerGetParentObject(timer));    
+    COLOR_CONTROL* pColorController = WdfObjectGet_COLOR_CONTROL(timer);
+    NTSTATUS status = SetFeatureColor(Device, pColorController->Colors[pColorController->RemainingColors]);
+    if (!NT_SUCCESS(status)) {
+        DPF_SOMETHING_FAILED(instance, something); }
+
+    if (pColorController->RemainingColors) {
+        pColorController->RemainingColors = pColorController->RemainingColors - 1;
+        
+        BOOLEAN timerInQueue = WdfTimerStart(timer, WDF_REL_TIMEOUT_IN_SEC(1));
+        NT_ASSERTMSG("Previous active timer overwritten", !timerInQueue);
+    }
+    else {
+        pColorController->RemainingColors = REMAINING_COLORS_COUNT;
+    }
+}
+
+NTSTATUS FakeBIST(_In_ WDFDEVICE Device) {
+/*++
+    
+ Routine Description: Performs a fake Built-In Self Test (BIST)
+
+--*/
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    WDFTIMER* pTimer = &WdfObjectGet_DEVICE_CONTEXT(Device)->FakeBISTTimer;
+    auto pColorController = WdfObjectGet_COLOR_CONTROL(*pTimer);
+
+    // FakeBIST is called in an arbitrary thread. If more than one thread is in
+    // FakeBIST at the same time, then the call to SetFeatureColor will result
+    // in an invalid taillight color sequence. Simply return that the device is
+    // busy using an invalid COLOR_CONTROL::RemainingColors value to signify
+    // that the hardware is free. Otherwise we can queue up the request or 
+    // block. Both of those mechanisms would require notifying that the
+    // hardware is busy and cancelation.
+    // 
+    // TODO: Maybe a function to get the last BIST test result and time.
+    if (pColorController->RemainingColors != REMAINING_COLORS_COUNT) {
+        return STATUS_DEVICE_BUSY;
+    }
+
+    InterlockedDecrement((LONG*) & pColorController->RemainingColors);
+
+    // Indicate that we are "starting" the test.
+    status = SetFeatureColor(Device, 0xFF);
+
+    pTimer = &WdfObjectGet_DEVICE_CONTEXT(Device)->FakeBISTTimer;
```

Comment:
Why are you re-assigning the `pTimer` pointer here? It was already assigned at the beginning of the function.